### ISSUE TITLE
Unescaped text entity -- raw()

### ISSUE THREAD

**Issue by marklister** [raquo/scala-dom-types#41]
Do you have an equivalent to scalatags' raw()?

**Comment by marklister** [comment]
The use case is explained in https://www.lihaoyi.com/scalatags/#Auto-escapingandunsanitizedInput

Basically if you want a text entity rendered without escaping "<" ">" etc in scalatags you enclose them in `raw()`.  I was hoping I could do the same in scala-dom-types

**Comment by raquo** [comment]
Hi Mark, my current policy is to not include such unsafe constructs in
scala dom types. SDT is designed to be consumed by higher level libraries,
each of which might have different opinions on how to handle these cases.
For example, in Lamjnar my policy is "the library will not provide this,
use native DOM methods".

In SDT terms, you can implement the `raw` tag or `innerHtml` property in
your own project similarly to how other tags and properties are implemented.

Alteratively, you can simply implement a `def raw(u safeHtmlContent:
String): Modifier[dom.Element]` method and expose it to your code/users.
The method would return a Modifier that sets the given element's
`innerHtml` property to the provided HTML string.

--
Nikita

**Comment by cornerman** [comment]
@raquo How about introducing an `UnsafeAttributes` trait? A consuming library can decide whether it wants to pull this in and if does not want to, the developer can just ignore it. It can also decide to make these definitions only accessible via an `unsafe`-prefix like `unsafe.innerHTML`.

I think, SDT is for typing html/svg/css elements and even if we think `innerHTML` might not be desired by default or is unsafe to do, it is part of the spec after all. It even might be totally fine in some consuming libraries. In outwatch (with a patched version of snabbdom :/), we can work with `innerHTML`. And it is useful if I want to e.g. render markdown as html.

**Comment by raquo** [comment]
Hrm, I guess we can do it this way for `innerHtml`, with the trait having a generic param for the types of values that this property should accept (e.g. not just string, but let the consuming libraries require some `RawHtml(string)` as input). Maybe this will do a better job of informing potential consuming libraries of the dangers of `innerHtml` than omitting it. Although, that would require the user to implement the `innerHtml` method since we won't be able to provide a codec, which... leaves no usefulness to the trait other than the single `innerHtml` name.

I'm not sure how ScalaTags' `raw` works, but it probably _appends_ to the parent's content rather than _replaces_ it like `innerHtml` does. That is definitely out of scope for SDT.

Related: https://github.com/raquo/scala-dom-types/pull/21

**Comment by cornerman** [comment]
Why did you close this issue? I still think, there is value in having something like `UnsafeHtmlKeys`. We could do it similar to what is done in https://github.com/raquo/scala-dom-types/pull/43/files. Or do you have another plan?

**Comment by raquo** [comment]
The issue itself is asking for functionality of `raw`, which is not exactly the same as `innerHTML` (appends, not replaces content), and has no direct alternative in the DOM. As such, to me it seems out of scope for SDT. We can still discuss `innerHTML` though. #43 Adds a link to this thread in Contributing FAQ for some more visibility.

I still think that it is not useful to have `innerHTML` in SDT. If we add that, what about `outerHTML` for example? It's an even crazier one, even less likely to be used. Do they go into the same trait? How useful would that trait be if we put a bunch of dangerous largely-undesired things in it? Someone would need to want all of those in their API to make use of that trait.

Ultimately we could split every controversial attribute in their own trait but that doesn't scale. I think it's reasonable to expect consuming libraries to implement a handful of controversial attributes like this. And if they do, there isn't much value to SDT providing unnecessarily generalized facilities for that.

In contrast to these fringe properties, `ComplexHtmlKeys` is supposed to have keys that every library would want to implement, but perhaps differently. TBH I'm still second guessing even adding `textContent` to it, I might actually remove it from there because I'm not sure how universally supported its behaviour – modifying the tree of elements – would be in consuming libraries.

---

My current opinion aside, after a quick search it seems that Outwatch doesn't have anything built in for innerHTML. Would Outwatch make use of innerHTML if it was available in SDT? If so, would you use a canonical SDT implementation that just sets the property, or override with your own?

**Comment by cornerman** [comment]
Yeah, a sophisticated implementation of `raw` is really not in the scope of this library. So closing the issue is probably right. 

> TBH I'm still second guessing even adding textContent to it, I might actually remove it from there because I'm not sure how universally supported its behaviour – modifying the tree of elements – would be in consuming libraries.

I think, it is the same as with `innerHTML` or `innerText`. Consuming libraries probably have other ways to append content to a node. The only use case I have for `innerText` and `textContent` right now is reading the value from a contenteditable, but that would be done via the dom element api anyhow.

> I still think that it is not useful to have innerHTML in SDT. If we add that, what about outerHTML for example? [...] Ultimately we could split every controversial attribute in their own trait but that doesn't scale

Good point. Adding `outerHTML` would indeed make such a trait useless for me, because I would not want that in my API :)

>  it seems that Outwatch doesn't have anything built in for innerHTML.

That is true right now, because snabbdom has an open issue to work with nodes that are edited outside of snabbdom patching with e.g. innerHTML (https://github.com/snabbdom/snabbdom/pull/347). So until this is merged, we cannot offer an API for this or would need to transform it into our vnode model. Though, in a web application we are writing, we use a fork of snabbdom and can work fine with `innerHTML`. We use it for rendering html that is parsed from markdown.

An open question is probably whether we would stick with the name `innerHTML` in our outwatch API or whether we would use a different name. The advantage of using the known html property name, is that user can discover it easily if they come from html. We would definitely override the type of it to be something like `case class UnsafeHtml(html: String)`.

**Comment by cornerman** [comment]
In the end, it really feels like a very opinionated API, because it really depends on the application whether you can support some of these properties. I just thought, it would be cool to have all of these attributes defined in SDT, so a consuming library could easily pull in all the definitions if it wanted to. But I do not have such a generic use-case right now and it feels like quite a hazzle to find an interface we are happy with.

So, I now tend to agree and will just define it downstream. We can still reiterate if there is a proper use-case :)
